File: Python/Problems/Problems/problem859.py
import heapq
import math
import sys

import numpy as np

from utils import print1DArray, print2DArray
from typing import List, Set, Optional
from collections import Counter

from .problem import Problem

class Problem859(Problem):
    def test(self) -> bool:
        s = "ab"
        goal = "ba"

        expected = True

        result = self.buddyStrings(s, goal)

        return result == expected

    def buddyStrings(self, s: str, goal: str) -> bool:
        n = len(s)
        if n != len(goal):
            return False
        diff = -1
        swapped = False
        multi = False
        present = [False] * 26
        for i in range(n):
            if not multi:
                id = ord(s[i]) - ord('a')
                if present[id]:
                    multi = True
                else:
                    present[id] = True
            if s[i] != goal[i]:
                if swapped:
                    return False
                if diff == -1:
                    diff = i
                else:
                    if s[i] != goal[diff] or s[diff] != goal[i]:
                        return False
                    swapped = True

        return swapped or (diff == -1 and multi)

File: C++/Problems/Problems/problem_859.cpp
#include "problem_859.h"

bool problem_859::test() {
    std::string s = "ab";
    std::string goal = "ba";

    bool expected = true;

    auto result = buddyStrings(s, goal);

    return result == expected;
}

bool problem_859::buddyStrings(std::string s, std::string goal) {
    int n = s.size();
    if (n != goal.size())
        return false;
    int diff = -1;
    bool swapped = false;
    bool multi = false;
    std::vector present(26, false);
    for (int i = 0; i < n; ++i)
    {
        if (!multi)
        {
            int id = s[i] - 'a';
            if (present[id])
                multi = true;
            else
                present[id] = true;
        }
        if (s[i] != goal[i])
        {
            if (swapped)
                return false;
            if (diff == -1)
                diff = i;
            else
            {
                if (s[i] != goal[diff] || s[diff] != goal[i])
                    return false;
                swapped = true;
            }

        }
    }

    return swapped || (diff == -1 && multi);
}
